File: app/src/main/java/com/github/anrimian/musicplayer/di/app/DbModule.java
package com.github.anrimian.musicplayer.di.app;

import android.content.Context;

import com.github.anrimian.musicplayer.data.database.AppDatabase;
import com.github.anrimian.musicplayer.data.database.DatabaseManager;
import com.github.anrimian.musicplayer.data.database.dao.albums.AlbumsDao;
import com.github.anrimian.musicplayer.data.database.dao.albums.AlbumsDaoWrapper;
import com.github.anrimian.musicplayer.data.database.dao.artist.ArtistsDao;
import com.github.anrimian.musicplayer.data.database.dao.artist.ArtistsDaoWrapper;
import com.github.anrimian.musicplayer.data.database.dao.compositions.CompositionsDao;
import com.github.anrimian.musicplayer.data.database.dao.compositions.CompositionsDaoWrapper;
import com.github.anrimian.musicplayer.data.database.dao.folders.FoldersDao;
import com.github.anrimian.musicplayer.data.database.dao.folders.FoldersDaoWrapper;
import com.github.anrimian.musicplayer.data.database.dao.genre.GenreDao;
import com.github.anrimian.musicplayer.data.database.dao.genre.GenresDaoWrapper;
import com.github.anrimian.musicplayer.data.database.dao.play_list.PlayListDao;
import com.github.anrimian.musicplayer.data.database.dao.play_list.PlayListsDaoWrapper;
import com.github.anrimian.musicplayer.data.database.dao.play_queue.PlayQueueDao;
import com.github.anrimian.musicplayer.data.database.dao.play_queue.PlayQueueDaoWrapper;

import javax.annotation.Nonnull;
import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

/**
 * Created on 20.11.2017.
 */

@Module
public class DbModule {

    @Provides
    @Nonnull
    @Singleton
    DatabaseManager provideDatabaseManager(Context context) {
        return new DatabaseManager(context);
    }

    @Provides
    @Nonnull
    @Singleton
    AppDatabase provideAppDatabase(DatabaseManager databaseManager) {
        return databaseManager.getAppDatabase();
    }

    @Provides
    @Nonnull
    @Singleton
    PlayQueueDao playQueueDao(AppDatabase appDatabase) {
        return appDatabase.playQueueDao();
    }

    @Provides
    @Nonnull
    @Singleton
    CompositionsDao compositionsDao(AppDatabase appDatabase) {
        return appDatabase.compositionsDao();
    }

    @Provides
    @Nonnull
    @Singleton
    PlayQueueDaoWrapper playQueueDaoWrapper(AppDatabase appDatabase, PlayQueueDao playQueueDao) {
        return new PlayQueueDaoWrapper(appDatabase, playQueueDao);
    }

    @Provides
    @Nonnull
    @Singleton
    ArtistsDao artistsDao(AppDatabase appDatabase) {
        return appDatabase.artistsDao();
    }

    @Provides
    @Nonnull
    @Singleton
    AlbumsDao albumssDao(AppDatabase appDatabase) {
        return appDatabase.albumsDao();
    }

    @Provides
    @Nonnull
    @Singleton
    GenreDao genreDao(AppDatabase appDatabase) {
        return appDatabase.genreDao();
    }

    @Provides
    @Nonnull
    @Singleton
    AlbumsDaoWrapper albumsDaoWrapper(AppDatabase appDatabase,
                                      AlbumsDao albumsDao,
                                      ArtistsDao artistsDao,
                                      ArtistsDaoWrapper artistsDaoWrapper) {
        return new AlbumsDaoWrapper(appDatabase, albumsDao, artistsDao, artistsDaoWrapper);
    }

    @Provides
    @Nonnull
    @Singleton
    ArtistsDaoWrapper artistsDaoWrapper(AppDatabase appDatabase,
                                        ArtistsDao artistsDao,
                                        AlbumsDao albumsDao) {
        return new ArtistsDaoWrapper(appDatabase, artistsDao, albumsDao);
    }

    @Provides
    @Nonnull
    @Singleton
    GenresDaoWrapper genresDaoWrapper(AppDatabase appDatabase,
                                      GenreDao genreDao,
                                      CompositionsDao compositionsDao) {
        return new GenresDaoWrapper(appDatabase, genreDao, compositionsDao);
    }

    @Provides
    @Nonnull
    @Singleton
    CompositionsDaoWrapper compositionsDaoWrapper(AppDatabase appDatabase,
                                                  ArtistsDao artistsDao,
                                                  CompositionsDao compositionsDao,
                                                  AlbumsDao albumsDao,
                                                  GenreDao genreDao,
                                                  FoldersDao foldersDao) {
        return new CompositionsDaoWrapper(appDatabase,
                artistsDao,
                compositionsDao,
                albumsDao,
                genreDao,
                foldersDao);
    }

    @Provides
    @Nonnull
    @Singleton
    FoldersDao foldersDao(AppDatabase appDatabase) {
        return appDatabase.foldersDao();
    }

    @Provides
    @Nonnull
    @Singleton
    FoldersDaoWrapper foldersDaoWrapper(AppDatabase appDatabase,
                                        FoldersDao foldersDao,
                                        CompositionsDaoWrapper compositionsDao) {
        return new FoldersDaoWrapper(appDatabase, foldersDao, compositionsDao);
    }

    @Provides
    @Nonnull
    @Singleton
    PlayListDao playListDao(AppDatabase appDatabase) {
        return appDatabase.playListDao();
    }

    @Provides
    @Nonnull
    @Singleton
    PlayListsDaoWrapper playListsDaoWrapper(PlayListDao playListDao,
                                            CompositionsDao compositionsDao,
                                            AppDatabase appDatabase) {
        return new PlayListsDaoWrapper(playListDao, compositionsDao, appDatabase);
    }
}

File: app/src/main/java/com/github/anrimian/musicplayer/di/app/DbModule.kt
package com.github.anrimian.musicplayer.di.app

import android.content.Context
import com.github.anrimian.musicplayer.data.database.ConfigsDatabase
import com.github.anrimian.musicplayer.data.database.DatabaseManager
import com.github.anrimian.musicplayer.data.database.LibraryDatabase
import com.github.anrimian.musicplayer.data.database.dao.albums.AlbumsDao
import com.github.anrimian.musicplayer.data.database.dao.albums.AlbumsDaoWrapper
import com.github.anrimian.musicplayer.data.database.dao.artist.ArtistsDao
import com.github.anrimian.musicplayer.data.database.dao.artist.ArtistsDaoWrapper
import com.github.anrimian.musicplayer.data.database.dao.compositions.CompositionsDao
import com.github.anrimian.musicplayer.data.database.dao.compositions.CompositionsDaoWrapper
import com.github.anrimian.musicplayer.data.database.dao.folders.FoldersDao
import com.github.anrimian.musicplayer.data.database.dao.folders.FoldersDaoWrapper
import com.github.anrimian.musicplayer.data.database.dao.genre.GenreDao
import com.github.anrimian.musicplayer.data.database.dao.genre.GenresDaoWrapper
import com.github.anrimian.musicplayer.data.database.dao.ignoredfolders.IgnoredFoldersDao
import com.github.anrimian.musicplayer.data.database.dao.play_queue.PlayQueueDao
import com.github.anrimian.musicplayer.data.database.dao.play_queue.PlayQueueDaoWrapper
import com.github.anrimian.musicplayer.data.database.dao.playlist.PlaylistDao
import com.github.anrimian.musicplayer.data.database.dao.playlist.PlaylistsDaoWrapper
import com.github.anrimian.musicplayer.data.storage.providers.music.SystemAudioCatalogProvider
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
class DbModule {

    @Provides
    @Singleton
    fun provideDatabaseManager(
        context: Context,
        systemAudioCatalogProvider: SystemAudioCatalogProvider,
    ) = DatabaseManager(context, systemAudioCatalogProvider)

    @Provides
    @Singleton
    fun provideAppDatabase(
        databaseManager: DatabaseManager
    ): LibraryDatabase = databaseManager.getLibraryDatabase()

    @Provides
    @Singleton
    fun playQueueDao(
        libraryDatabase: LibraryDatabase
    ): PlayQueueDao = libraryDatabase.playQueueDao()

    @Provides
    @Singleton
    fun compositionsDao(
        libraryDatabase: LibraryDatabase
    ): CompositionsDao = libraryDatabase.compositionsDao()

    @Provides
    @Singleton
    fun playQueueDaoWrapper(
        libraryDatabase: LibraryDatabase,
        playQueueDao: PlayQueueDao
    ) = PlayQueueDaoWrapper(libraryDatabase, playQueueDao)

    @Provides
    @Singleton
    fun artistsDao(
        libraryDatabase: LibraryDatabase
    ): ArtistsDao = libraryDatabase.artistsDao()

    @Provides
    @Singleton
    fun albumsDao(
        libraryDatabase: LibraryDatabase
    ): AlbumsDao = libraryDatabase.albumsDao()

    @Provides
    @Singleton
    fun genreDao(
        libraryDatabase: LibraryDatabase
    ): GenreDao = libraryDatabase.genreDao()

    @Provides
    @Singleton
    fun albumsDaoWrapper(
        libraryDatabase: LibraryDatabase,
        albumsDao: AlbumsDao,
        artistsDao: ArtistsDao,
        artistsDaoWrapper: ArtistsDaoWrapper
    ) = AlbumsDaoWrapper(libraryDatabase, albumsDao, artistsDao, artistsDaoWrapper)

    @Provides
    @Singleton
    fun artistsDaoWrapper(
        libraryDatabase: LibraryDatabase,
        artistsDao: ArtistsDao,
        albumsDao: AlbumsDao
    ) = ArtistsDaoWrapper(libraryDatabase, artistsDao, albumsDao)

    @Provides
    @Singleton
    fun genresDaoWrapper(
        libraryDatabase: LibraryDatabase,
        genreDao: GenreDao,
        compositionsDao: CompositionsDao
    ) = GenresDaoWrapper(libraryDatabase, genreDao, compositionsDao)

    @Provides
    @Singleton
    fun compositionsDaoWrapper(
        libraryDatabase: LibraryDatabase,
        artistsDao: ArtistsDao,
        artistsDaoWrapper: ArtistsDaoWrapper,
        compositionsDao: CompositionsDao,
        albumsDao: AlbumsDao,
        genreDao: GenreDao,
        foldersDao: FoldersDao
    ) = CompositionsDaoWrapper(
        libraryDatabase,
        artistsDao,
        artistsDaoWrapper,
        compositionsDao,
        albumsDao,
        genreDao,
        foldersDao
    )

    @Provides
    @Singleton
    fun foldersDao(
        libraryDatabase: LibraryDatabase
    ): FoldersDao = libraryDatabase.foldersDao()

    @Provides
    @Singleton
    fun foldersDaoWrapper(
        libraryDatabase: LibraryDatabase,
        foldersDao: FoldersDao,
        compositionsDao: CompositionsDaoWrapper,
    ) = FoldersDaoWrapper(libraryDatabase, foldersDao, compositionsDao)

    @Provides
    @Singleton
    fun playListDao(
        libraryDatabase: LibraryDatabase
    ): PlaylistDao = libraryDatabase.playListDao()

    @Provides
    @Singleton
    fun playListsDaoWrapper(
        playListDao: PlaylistDao,
        compositionsDao: CompositionsDao,
        libraryDatabase: LibraryDatabase
    ) = PlaylistsDaoWrapper(playListDao, compositionsDao, libraryDatabase)

    @Provides
    @Singleton
    fun configsDatabase(
        databaseManager: DatabaseManager
    ): ConfigsDatabase = databaseManager.getConfigsDatabase()

    @Provides
    @Singleton
    fun ignoredFoldersDao(
        configsDatabase: ConfigsDatabase
    ): IgnoredFoldersDao = configsDatabase.ignoredFoldersDao()
}
